tensorflow/python/ops/array_ops.py

```diff
@@ -557,6 +557,14 @@ def shape_v2(input, out_type=dtypes.int32, name=None):
 
   >>> a.shape
   TensorShape([None, None, 10])
+  
+  However, when defining custom layers and models that will be run in graph mode
+  at some point, prefer `tf.shape(x)` over `x.shape`. `x.shape` is the static shape
+  of `x` and usually evaluates to `None` in the first dimension during graph
+  construction (to represent the as yet unknown batch size). This can cause problems in
+  function calls like `tf.zeros(x.shape[0])` which don't support `None` values.
+  `tf.shape(x)` on the other hand gives the dynamic shape of `x` which isn't
+  evaluated until training/predicting begins where the full shape of `x`  is known.
```

Comment:
The example above this says exactly this. Note the difference between `tf.shape(a)` and `a.shape` above.

Reply:
I find the information above inactionable. Even reading it now already knowing what it's trying to tell me, I have a hard time connecting the difference in behavior between `tf.shape(x)` and `x.shape` to the decision when to use which. The docs could be clearer on that point in my opinion. Hence the hint to prefer `tf.shape(x)` over `x.shape` when running in graph mode. Maybe we can shorten it?

Reply:
Agree with shortening.

Reply:
Do you have something in mind or should I have a go?

Reply:
Better you have a go. The relevant parts that should be included are that `tf.shape(x)` computes the dynamic shape whereas `x.shape` is the static shape known at the attribute access location. Then maybe one short phrase of where this would be different, but not too long

Reply:
Gave it a shot.

Reply:
Thanks Janosh. I read one SO answer where a googler working in TF Team answered it with great clarity. May be it will help you update and bring out the difference between tf.shape(x) and x.shape. [Here](https://stackoverflow.com/questions/37096225/how-to-understand-static-shape-and-dynamic-shape-in-tensorflow) is the link to that SO answer.